caas/kubernetes/provider/rbac.go

```diff
@@ -0,0 +1,463 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"github.com/juju/errors"
+	core "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/juju/juju/caas"
+)
+
+func (k *kubernetesClient) getRBACLabels(appName string) map[string]string {
+	return map[string]string{
+		labelApplication: appName,
+		labelModel:       k.namespace,
+	}
+}
+
+func (k *kubernetesClient) ensureServiceAccountForApp(
+	appName string, caasSpec *caas.ServiceAccountSpec,
+) (cleanups []func(), err error) {
+	labels := k.getRBACLabels(appName)
+	saSpec := &core.ServiceAccount{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      caasSpec.Name,
+			Namespace: k.namespace,
+			Labels:    labels,
+		},
+		AutomountServiceAccountToken: caasSpec.AutomountServiceAccountToken,
+	}
+	// ensure service account;
+	sa, err := k.updateServiceAccountForApp(appName, saSpec)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return cleanups, errors.Trace(err)
+		}
+		if sa, err = k.createServiceAccount(saSpec); err != nil {
+			return cleanups, errors.Trace(err)
+		}
+		cleanups = append(cleanups, func() { k.deleteServiceAccount(sa.GetName()) })
+	}
+
+	roleSpec := caasSpec.Capabilities.Role
+	var roleRef rbacv1.RoleRef
+	switch roleSpec.Type {
+	case caas.Role:
+		var r *rbacv1.Role
+		if len(roleSpec.Rules) == 0 {
+			// no rules was specified, reference to an existing Role.
+			r, err = k.getRole(roleSpec.Name)
+			if err != nil {
+				return cleanups, errors.Trace(err)
+			}
+		} else {
+			// create or update Role.
+			r, err = k.ensureRole(&rbacv1.Role{
```

Comment:
This seems like the relevant code to Mike's question: this will use an existing role with the right name and overwrite its rules, I think? 
If someone deploys two charms that define roles with the same name but different rules, will they clash and whichever is deployed last wins? This seems like it would cause problems, since the loser wouldn't be able to perform the operations it expects to.

A better option could be to get the existing role and check that the rules match for both, otherwise to error? Although that seems like it could cause problems on upgrade if the role needed to change.

I might be misunderstanding - is the namespace application-specific or  model-wide here?

Reply:
yeah, raise a good point!
fixing now